UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs

```diff
@@ -170,4 +169,13 @@ private void OnPitchDetected(int midiPitch)
         PitchEvent pitchEvent = new PitchEvent(midiNoteMedian);
         pitchEventStream.OnNext(pitchEvent);
     }
+
+    private void OnNoPitchDetected()
+    {
+        // no tone detected.
+        // Subscribers (such as the PlayerNoteRecorder) must know that the singing ended.
+        // Therefore, a midi pitch of 0 is interpreted as "no singing".
+        pitchRecordHistory.Clear();
```

Comment:
IMO a "no pitch" of 0 should not have priority over other pitch values. Clearing the history makes it brittle again.

Imagine a pitch detection stream of midi note [60, 60, 60, 0, 60]. In this sequence the 0 would be irrelevant. The median is still 60. But when clearing the history, the 0 will always show up.

Instead, I would implement OnNoPitchDetected() as OnPitchDetected(0);

Reply:
I disagree. The purpose of the pitchRecordHistory is to filter out short-time noise which is being detected as pitches. Writing 0s in there completely breaks this mechanism and does not improve pitch detection.
The purpose of the pitch detection code is to detect if a pitch was sung, and if yes, return that pitch.  Returning a new pitchEvent of 0 is already a hack which eventually should be removed from the CamdAudioSamplesAnalyzer . 
Without this pull request, there were bunch of cases where no pitchEvent was thrown at all, thus no result for the grading was available for some times and the pitch text display in the record settings occasionally didn't update properly.
The pitchRecordHistory.Clear(); was added because after the player stopped singing a word, we know that very likely a new syllable will be sung and thus the assumption that pitch will most likely not change doesn't fit there any more.